src/lib/apply-patch/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+                    // this may be a whitespace-insensitive patch, and we don't want to modify any lines that aren't marked `-`
+                    patched_content.push_str(orig_lines.next().unwrap());
+                    patched_content.push('\n');
+                }
+                Line::Add(s) => {
+                    patched_content.push_str(s);
+                    patched_content.push('\n');
+                }
+                Line::Remove(_) => {
+                    orig_lines.next().unwrap();
+                }
+            }
+        }
+    }
+
+    fn find_fixed_range(
+        &self,
+        original_content: &str,
+        hunk: &Hunk,
+        expected_lines: &[&str],
+    ) -> Result<Range<usize>, MismatchError> {
+        let mut lines = original_content.match_indices('\n');
+        // line numbers are all 1-indexed - this is specifically the parsed number
+        let start_line = hunk.old_range.start as usize - 1;
+        // line n starts one past the n-1th newline
+        let start = if start_line == 0 {
+            0
+        } else {
+            lines
+                .nth(start_line - 1)
+                .ok_or_else(|| MismatchError::NotEnoughLines {
+                    expected: start_line - 1,
+                    found: original_content.lines().count(),
+                })?
+                .0
+                + 1
+        };
+        // start..end should be the byte range in `content` to be patched
+        let end = lines
+            .nth(expected_lines.len() - 1)
+            .ok_or_else(|| MismatchError::NotEnoughLines {
+                expected: expected_lines.len() - 1,
+                found: original_content.lines().count(),
+            })?
+            .0
+            + 1;
+        check_equal_range(
+            &original_content[start..end],
+            expected_lines,
+            self.ignore_whitespace,
+            start_line,
+        )?;
+        Ok(start..end)
+    }
+}
+
+fn is_patch_coherent(patch: &Patch) -> bool {
+    // each patch file should be sorted and non-overlapping
+    patch
+        .hunks
+        .iter()
+        .zip(&patch.hunks[1..])
+        .all(|(h1, h2)| h1.old_range.start + h1.old_range.count < h2.old_range.start)
+}
+fn check_equal_range(
+    content: &str,
+    lines: &[&str],
+    ignore_whitespace: bool,
+    context_line_number: usize,
+) -> Result<(), MismatchError> {
+    for (i, (from_content, &from_patch)) in content.lines().zip(lines).enumerate() {
+        if ignore_whitespace {
+            if !from_content
+                .chars()
+                .filter(|ch| !ch.is_whitespace())
+                .eq(from_patch.chars().filter(|ch| !ch.is_whitespace()))
+            {
+                return Err(MismatchError::LineMismatch {
+                    from_content: from_content.to_string(),
+                    from_patch: from_patch.to_string(),
+                    whitespace_insensitive: ignore_whitespace,
```

Comment:
We now have three ways in the code of expressing a similar idea:
- ignore_whitespace
- whitespace_insensitive
- exact_whitespace

Could we pick one?

Reply:
I would prefer to keep `ignore_whitespace` and `whitespace_insensitive` in the API - one is an instruction to the system, one is the state of a match that produced an error. The name of the internal variable can be changed to match one of the two, though.